src/read_logs_from.rs

```diff
@@ -0,0 +1,148 @@
+    pub shutdown: ShutdownHandle,
+    /// Path to read logs from.
+    pub path: Utf8PathBuf,
+}
+
+impl ReadLogsFrom {
+    /// Read logs from the given path and output them to stdout.
+    #[instrument(skip_all, name = "read-logs", level = "debug", fields(path = %self.path))]
+    pub async fn run(mut self) -> miette::Result<()> {
+        let mut backoff = ExponentialBackoff {
+            max_elapsed_time: None,
+            max_interval: Duration::from_secs(1),
+            ..Default::default()
+        };
+        while let Some(duration) = backoff.next_backoff() {
+            match self.run_inner().await {
+                Ok(()) => {
+                    // Graceful exit.
+                    break;
+                }
+                Err(err) => {
+                    // These errors are often like "the file doesn't exist yet" so we don't want
+                    // them to be noisy.
+                    tracing::debug!("{err:?}");
+                }
+            }
+
+            tracing::debug!("Waiting {duration:?} before retrying");
+            tokio::time::sleep(duration).await;
+        }
+
+        Ok(())
+    }
+
+    async fn run_inner(&mut self) -> miette::Result<()> {
+        loop {
+            tokio::select! {
+                result = Self::read(&self.path) => {
+                    result?;
+                }
+                _ = self.shutdown.on_shutdown_requested() => {
+                    // Graceful exit.
+                    break;
+                }
+                else => {
+                    // Graceful exit.
+                    break;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn read(path: &Utf8Path) -> miette::Result<()> {
+        let file = File::open(&path).await.into_diagnostic()?;
+        let mut metadata = file.metadata().await.into_diagnostic()?;
+        let mut size = metadata.len();
+        let mut reader = BufReader::new(file);
+
+        if size > MAX_BYTES_PRINT_FROM_END {
+            tracing::debug!("Log file too big, skipping to end");
+            reader
+                .seek(SeekFrom::End(
+                    -MAX_BYTES_PRINT_FROM_END
+                        .try_conv::<i64>()
+                        .expect("Constant is not bigger than i64::MAX"),
+                ))
+                .await
+                .into_diagnostic()?;
+        }
+
+        let mut lines = reader.lines();
+
+        let mut backoff = ExponentialBackoff {
+            max_elapsed_time: None,
+            max_interval: Duration::from_millis(1000),
+            ..Default::default()
+        };
+
+        let mut stdout = tokio::io::stdout();
+
+        while let Some(duration) = backoff.next_backoff() {
+            while let Some(line) = lines.next_line().await.into_diagnostic()? {
+                // TODO: Lock stdout here and for ghci output.
+                let _ = stdout.write_all(line.as_bytes()).await;
+                let _ = stdout.write_all(b"\n").await;
```

Comment:
This is a pretty big "TODO". `std::io::Stdout` is globally synchronized, but [`tokio::io::Stdout`](https://docs.rs/tokio/latest/tokio/io/struct.Stdout.html) is not, so this could lead to interleaved output.

AFAICT there's no easy solution for this. The tokio devs suggested creating a task to manage stdout and sending it messages (strings to write) over a channel, which is reasonable enough, but then I'll have to write an adapter to implement [`std::io::Write`](https://doc.rust-lang.org/stable/std/io/trait.Write.html) by sending messages to an async channel.